Sources/PointFree/Episode/Show.swift

```diff
@@ -45,29 +48,86 @@ let episodeResponse =
+private func episodePageData(
+  currentUser: User?,
+  collectionSlug: Episode.Collection.Slug?,
+  episode: Episode,
+  permission: EpisodePermission,
+  subscriberState: SubscriberState
+) -> Either<EpisodePageData, NewEpisodePageData> {
+  if currentUser?.isAdmin == .some(true) {
+    let context: NewEpisodePageData.Context
+    if let collection = Episode.Collection.all.first(where: { $0.slug == collectionSlug }) {
+      context = .collection(collection)
+    } else {
+      context = .direct(
+        previousEpisode: Current.episodes().first(where: { $0.sequence == episode.sequence - 1 }),
+        nextEpisode: Current.episodes().first(where: { $0.sequence == episode.sequence + 1 })
+      )
+    }
+
+    return .right(NewEpisodePageData(
+      context: context,
+      date: Current.date,
+      episode: episode,
+      permission: permission,
+      subscriberState: subscriberState,
+      user: currentUser
+    ))
+  } else {
+    return .left(EpisodePageData(
+      permission: permission,
+      user: currentUser,
+      subscriberState: subscriberState,
+      episode: episode,
+      previousEpisodes: episode.previousEpisodes,
+      date: Current.date
+    ))
+  }
+}
+
+private func newOrLegacyEpisodeView(data: Either<EpisodePageData, NewEpisodePageData>) -> Node {
+  switch data {
+  case let .left(data): return episodeView(episodePageData: data)
+  case let .right(data): return newEpisodePageView(episodePageData: data)
+  }
+}
+
+func useCreditResponse<Z>(
+  conn: Conn<StatusLineOpen, T5<Either<String, Episode.Id>, User?, SubscriberState, Route?, Z>>
+) -> IO<Conn<ResponseEnded, Data>> {
+  conn
+    |> (fetchEpisodeForParam
     <<< validateUserEpisodePermission
-    <| applyCreditMiddleware
-
-let fetchEpisodeForParam
-  : MT<
-  Tuple4<Either<String, Episode.Id>, User?, SubscriberState, Route?>,
-  Tuple4<Episode, User?, SubscriberState, Route?>
-  >
-  = filterMap(
-    over1(episode(forParam:)) >>> require1 >>> pure,
-    or: writeStatus(.notFound) >=> respond(lower >>> episodeNotFoundView)
-)
+    <| applyCreditMiddleware)
+}
 
-private let validateUserEpisodePermission
-  : MT<
-  Tuple4<Episode, User?, SubscriberState, Route?>,
-  Tuple5<EpisodePermission, Episode, User, SubscriberState, Route?>
-  >
-  = { userEpisodePermission >=> $0 }
+private func fetchEpisodeForParam<Z>(
+  middleware: @escaping M<T5<Episode, User?, SubscriberState, Route?, Z>>
+) -> M<T5<Either<String, Episode.Id>, User?, SubscriberState, Route?, Z>> {
+  middleware
+    |> filterMap(
+      over1(episode(forParam:)) >>> require1 >>> pure,
+      or: episodeNotFoundResponse
+  )
+}
+
+private func episodeNotFoundResponse<Z>(
+  conn: Conn<StatusLineOpen, T5<Either<String, Episode.Id>, User?, SubscriberState, Route?, Z>>
+) -> IO<Conn<ResponseEnded, Data>> {
+  conn
+    |> writeStatus(.notFound)
+    >=> respond { episodeNotFoundView(user: get2($0), subscriberState: get3($0), route: get4($0)) }
+}
+
+private func validateUserEpisodePermission<Z>(
```

Comment:
and here

all of that allows us to reuse those middlewares in a few situations.

for example, the middleware that uses a credit was being run off of some of these, which required us to pass along a collection slug even though it doesn't care about it. by generalizing it a bit I was able to make it take only the data it needed.